src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -289,6 +289,15 @@ protected NodeContext GetNode(string msbuildLocation, string commandLineArgs, in
                     CommunicationsUtilities.Trace("Successfully connected to created node {0} which is PID {1}", nodeId, msbuildProcess.Id);
                     return new NodeContext(nodeId, msbuildProcess, nodeStream, factory, terminateNode);
                 }
+
+                if (msbuildProcess.HasExited)
+                {
+                    CommunicationsUtilities.Trace($"Could not connect to node with PID {msbuildProcess.Id}; it has exited. This can indicate a crash at startup");
```

Comment:
BTW, it's small beans, but all these interpolated strings are creating strings even if you're not tracing (the 99% case). An alternative is to wrap the call with a check for tracing, another option is to pass arguments for the logging code to do string.format.

Reply:
You're right. I erroneously convinced myself that this was fine because "we're already in an error case" but spaced that inside the call we might (and normally will) do nothing.